indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/CostBasedAutoScaler.java

```diff
@@ -0,0 +1,310 @@
+import org.apache.druid.indexing.seekablestream.supervisor.SeekableStreamSupervisor;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.concurrent.Execs;
+import org.apache.druid.java.util.emitter.EmittingLogger;
+import org.apache.druid.java.util.emitter.service.ServiceEmitter;
+import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
+import org.apache.druid.query.DruidMetrics;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Cost-based auto-scaler for seekable stream supervisors.
+ * Uses a cost function combining lag and idle time metrics to determine optimal task counts.
+ * Task counts are selected from predefined values (not arbitrary factors).
+ * Scale-up happens incrementally, scale-down only during task rollover.
+ */
+public class CostBasedAutoScaler implements SupervisorTaskAutoScaler
+{
+  private static final EmittingLogger log = new EmittingLogger(CostBasedAutoScaler.class);
+
+  private static final int SCALE_FACTOR_DISCRETE_DISTANCE = 2;
+
+  private static final Map<Integer, List<Integer>> FACTORS_CACHE = new HashMap<>();
+
+  private final String dataSource;
+  /**
+   * Atomic reference to CostMetrics object. All operations must be performed
+   * with sequentially consistent semantics (volatile reads/writes).
+   * However, it may be fine-tuned with acquire/release semantics,
+   * but requires careful reasoning about correctness.
+   */
+  private final AtomicReference<CostMetrics> currentMetrics;
+  private final ScheduledExecutorService metricsCollectionExec;
+  private final ScheduledExecutorService scalingDecisionExec;
+  private final SupervisorSpec spec;
+  private final SeekableStreamSupervisor supervisor;
+  private final CostBasedAutoScalerConfig config;
+  private final ServiceEmitter emitter;
+  private final ServiceMetricEvent.Builder metricBuilder;
+  private final WeightedCostFunction costFunction;
+
+  public CostBasedAutoScaler(
+      SeekableStreamSupervisor supervisor,
+      String dataSource,
+      CostBasedAutoScalerConfig config,
+      SupervisorSpec spec,
+      ServiceEmitter emitter
+  )
+  {
+    this.dataSource = dataSource;
+    this.config = config;
+    this.spec = spec;
+    this.supervisor = supervisor;
+    this.emitter = emitter;
+
+    final String supervisorId = StringUtils.format("Supervisor-%s", dataSource);
+
+    this.currentMetrics = new AtomicReference<>(null);
+    this.costFunction = new WeightedCostFunction();
+
+    this.metricsCollectionExec = Execs.scheduledSingleThreaded(
+        StringUtils.encodeForFormat(supervisorId) + "-CostBasedMetrics-%d"
+    );
+    this.scalingDecisionExec = Execs.scheduledSingleThreaded(
```

Comment:
Do we really need two execs? We could have just 1 thread which runs at a certain period and first collects metrics, then performs scaling computations using those metrics.

Reply:
Totally agree.